pkg/ddc/juicefs/dataset.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2021 The Fluid Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package juicefs
+
+import (
+	"context"
+	"reflect"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/util/retry"
+
+	datav1alpha1 "github.com/fluid-cloudnative/fluid/api/v1alpha1"
+	"github.com/fluid-cloudnative/fluid/pkg/common"
+	"github.com/fluid-cloudnative/fluid/pkg/utils"
+)
+
+// UpdateDatasetStatus updates the status of the dataset
+func (j *JuiceFSEngine) UpdateDatasetStatus(phase datav1alpha1.DatasetPhase) (err error) {
+	// 1. update the runtime status
+	runtime, err := j.getRuntime()
+	if err != nil {
+		return err
+	}
+
+	// 2. update the dataset status
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		dataset, err := utils.GetDataset(j.Client, j.name, j.namespace)
+		if err != nil {
+			return err
+		}
+		datasetToUpdate := dataset.DeepCopy()
+		var cond datav1alpha1.DatasetCondition
+
+		switch phase {
+		case datav1alpha1.BoundDatasetPhase:
+			if len(datasetToUpdate.Status.Mounts) == 0 {
+				datasetToUpdate.Status.Mounts = datasetToUpdate.Spec.Mounts
+			}
+			cond = utils.NewDatasetCondition(datav1alpha1.DatasetReady, datav1alpha1.DatasetReadyReason,
+				"The ddc runtime is ready.",
+				corev1.ConditionTrue)
+		case datav1alpha1.FailedDatasetPhase:
+			cond = utils.NewDatasetCondition(datav1alpha1.DatasetReady, datav1alpha1.DatasetReadyReason,
+				"The ddc runtime is not ready.",
+				corev1.ConditionFalse)
+		default:
+			cond = utils.NewDatasetCondition(datav1alpha1.DatasetReady, datav1alpha1.DatasetReadyReason,
+				"The ddc runtime is unknown.",
+				corev1.ConditionFalse)
+		}
+
+		datasetToUpdate.Status.Phase = phase
+		datasetToUpdate.Status.Conditions = utils.UpdateDatasetCondition(datasetToUpdate.Status.Conditions,
+			cond)
+
+		datasetToUpdate.Status.CacheStates = runtime.Status.CacheStates
+
+		if !reflect.DeepEqual(dataset.Status, datasetToUpdate.Status) {
+			err = j.Client.Status().Update(context.TODO(), datasetToUpdate)
+			if err != nil {
+				j.Log.Error(err, "Update dataset")
```

Comment:
If the error message contains `NamespacedName` information, it is more convenient to locate problems when there are multiple instances online